### PR TITLE
Basic UI tests -- initial one for quick selector add step

### PR DESCRIPTION
**Pull request by jywarren**
re #1000

### CODE REVIEW

**Comment by jywarren** [comment]
```
Failures:
1) Default sequencer HTML adds a step from the quick selector
  Message:
    Expected 0 to be 1.
  Stack:
    Error: Expected 0 to be 1.
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/publiclab/image-sequencer/test/ui/spec/defaultHtmlSequencerUi.spec.js:44:31)
        at <Jasmine>
  Message:
    Expected 0 to be 2.
  Stack:
    Error: Expected 0 to be 2.
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/publiclab/image-sequencer/test/ui/spec/defaultHtmlSequencerUi.spec.js:46:31)
        at <Jasmine>
```

Needs clicking `Add Step`!

**Comment by jywarren** [comment]
Actually no, the quick selector doesn't need an Add Step click. So maybe the lib is not fully set up here yet...

**Comment by jywarren** [comment]
We need a fixture to have an HTML page to work on.

File: test/ui/spec/defaultHtmlSequencerUi.spec.js [hunk 1]
```diff
@@ -4,6 +4,10 @@ describe('Default sequencer HTML', function() {
   var sequencer = require('../../../src/ImageSequencer')();
```
Comment (jywarren) [inline test/ui/spec/defaultHtmlSequencerUi.spec.js]: So, here it looks like this is already done in the default UI code. So maybe we need to split this into another file to have one running purely on the fixture file.

**Comment by jywarren** [comment]
Moved tests into https://github.com/publiclab/image-sequencer/blob/ui-tests/test/ui/spec/basicUi.spec.js to avoid collision with the pure testing of the UI library, whereas `basicUi.spec.js` uses an HTML fixture to simulate a page.

**Comment by jywarren** [comment]
Ok so ReferenceError: loadFixtures is not defined

We prob need extra Jasmine modules for fixtures as per https://github.com/jywarren/pxlqst/blob/15887acea6bc297f1acd4785ad4f0ce400abfef4/package.json#L25

**Comment by jywarren** [comment]
Hmm. we do have that. Checking...

**Comment by jywarren** [comment]
I'm seeing a lot of this error:

```
events.js:174
      throw er; // Unhandled 'error' event
      ^
ReferenceError: window is not defined
    at Object.extraManipulation (/home/travis/build/publiclab/image-sequencer/src/modules/Average/Module.js:9:722)
```

**Comment by jywarren** [comment]
However i think the failing one is this, which says it needs jQuery!

```
Failures:
1) Default sequencer HTML adds a step from the quick selector
```

**Comment by jywarren** [comment]
Ah, progress! Ok, modeling on: https://github.com/jywarren/pxlqst/blob/master/spec/javascripts/pxlqst_spec.js

```
Failures:
1) Default sequencer HTML adds a step from the quick selector
  Message:
    Expected 0 to be 1.
  Stack:
    Error: Expected 0 to be 1.
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/publiclab/image-sequencer/test/ui/spec/basicUi.spec.js:8:31)
        at <Jasmine>
  Message:
    TypeError: Cannot read property 'src' of undefined
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/publiclab/image-sequencer/test/ui/spec/basicUi.spec.js:9:35)
        at <Jasmine>
        at runCallback (timers.js:810:20)
        at tryOnImmediate (timers.js:768:5)
        at processImmediate [as _immediateCallback] (timers.js:745:5)
Suite error: Default sequencer HTML
  Message:
    ReferenceError: loadFixtures is not defined
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/publiclab/image-sequencer/test/ui/spec/basicUi.spec.js:4:5)
        at <Jasmine>
        at runCallback (timers.js:810:20)
        at tryOnImmediate (timers.js:768:5)
        at processImmediate [as _immediateCallback] (timers.js:745:5)
24 specs, 2 failures
```

**Comment by jywarren** [comment]
I'm afraid full HTML UI tests may not be possible without the /dist/ files built and included in the library. I think this is a good reason to start tracking them again, though definitely keeping them .gitignored. cc @publiclab/is-reviewers for input?

**Comment by jywarren** [comment]
or is there another way to run the tests? Like, could we add a `grunt build` step to the test suite workflow?

**Comment by jywarren** [comment]
I tried adding `grunt browserify &&` to the `npm test` command! 🤞

**Comment by jywarren** [comment]
OK cool, it built! Maybe... I need to ref the resulting files now?

```
> image-sequencer@3.5.1 test-ui /home/travis/build/publiclab/image-sequencer
> grunt browserify && node node_modules/jasmine/bin/jasmine test/ui/spec/*.js
Running "browserify:core" (browserify) task
>> Bundle dist/image-sequencer.js created.
Running "browserify:ui" (browserify) task
>> Bundle dist/image-sequencer-ui.js created.
Running "browserify:prodcore" (browserify) task
>> Bundle dist/image-sequencer.brow.js created.
Running "browserify:produi" (browserify) task
>> Bundle dist/image-sequencer-ui.brow.js created.
```

Hmm. No, they are already referenced.

**Comment by jywarren** [comment]
Maybe we need to explicitly assign $ to window.$?

**Comment by jywarren** [comment]
Ah, whoops, maybe we need https://github.com/velesin/jasmine-jquery not jasmine-ajax?

File: test/ui/spec/support/jasmine.json [hunk 2]
```diff
@@ -4,7 +4,12 @@
     "**/*.[sS]pec.js"
   ],
   "helpers": [
-    "helpers/**/*.js"
+    "helpers/**/*.js",
+    "node_modules/jquery/dist/jquery.min.js",
```
Comment (jywarren) [inline test/ui/spec/support/jasmine.json]: Aha! These are supposed to be relative to the spec directory! https://github.com/jasmine/jasmine/issues/1109#issuecomment-239951622

**Comment by jywarren** [comment]
Ooh! Got past the $ error! Now it says: 

```

    TypeError: Cannot read property 'src' of undefined
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/publiclab/image-sequencer/test/ui/spec/basicUi.spec.js:9:35)
```

and
```
    ReferenceError: loadFixtures is not defined
```

So i think we have jquery now, but maybe not jasmine-jquery and `loadFixtures()`?

**Comment by jywarren** [comment]
Getting closer!
```
Default sequencer HTML
   - adds a step from the quick selector......X
     Error: Fixture could not be loaded: spec/javascripts/fixtures/index.html (status: error, message: undefined) in file:///home/travis/build/publiclab/image-sequencer/node_modules/jasmine-jquery/lib/jasmine-jquery.js (line 139) (1)
     Expected 0 to be 1. (2)
     TypeError: undefined is not an object (evaluating '$('.step:first img')[0].src') in file:///home/travis/build/publiclab/image-sequencer/test/ui/spec/basicUi.spec.js (line 9) (3)
 HTML Types Mapping Function
   - encountered a declaration exception......X
     ReferenceError: Can't find variable: require in file:///home/travis/build/publiclab/image-sequencer/test/ui/spec/mapHtmlTypes.spec.js (line 3) (1)
2 specs in 0.025s.
>> 2 failures
(node:8356) [DEP0013] DeprecationWarning: Calling an asynchronous function without callback is deprecated.
Warning: Task "jasmine:imageSequencer" failed. Use --force to continue.
```

**Comment by jywarren** [comment]
OK, so, managed to get the fixtures loading. But, I can't seem to get the libraries themselves to load. I'm seeing:

```js
     ReferenceError: Can't find variable: initializeSequencerUi in file:///home/warren/sites/image-sequencer/test/ui/spec/basicUi.spec.js (line 6) (1)
```

It's odd. I can't get ImageSequencer to load either, despite it's inclusion in the Gruntfile. I also tried using a `<script>` tag in the fixture, but that didn't work either. 

Basing this on https://github.com/publiclab/PublicLab.Editor/blob/master/spec/javascripts/editor_spec.js and https://github.com/publiclab/PublicLab.Editor/blob/master/Gruntfile.js

@publiclab/is-reviewers any suggestions?

**Comment by jywarren** [comment]
Hi @ananyaarun -- i'm struggling a bit getting these tests to run. I was wondering if you'd be able to look at it since you so recently got this running in LEL? Have I overlooked something simple?

**Comment by ananyaarun** [comment]
Hey @jywarren , sure I'll give it a look.

**Comment by jywarren** [comment]
@Divy123 @HarshKhandeparkar @rexagod any ideas here? :-/

**Comment by jywarren** [comment]
Referencing this similar PR from another project: https://github.com/publiclab/leaflet-environmental-layers/pull/247

**Comment by jywarren** [comment]
Could it relate to these includes in the file @sagarpreet-chadha set up over in `leaflet-environmental-layers`? https://github.com/publiclab/leaflet-environmental-layers/pull/206/files#diff-d798dca1a53c3a3773cd503064af3093R15-R39

**Comment by vaibhavmatta** [comment]
@jywarren May I help here?

**Comment by jywarren** [comment]
@vaibhavmatta i'd love help, sorry I didn't see this! I'm just trying to get the most basic browser-based tests running that we can build on!

**Comment by vaibhavmatta** [comment]
@jywarren I think basic browser-based tests won't work so efficiently. But if you say, I would love to work on them as well.! Thanks 🙂

**Comment by jywarren** [comment]
Well, we really need browser-based tests of the UI buttons, as they keep
breaking. We won't do the entire suite of tests here, but just a few
critical ones to ensure the HTML UI is stable. Thanks!!!

**Comment by vaibhavmatta** [comment]
Alright.! I got you 🙂

**Comment by SidharthBansal** [comment]
@keshav234156 is interested in doing this

**Comment by jywarren** [comment]
That would be amazing.

**Comment by keshav234156** [comment]
@jywarren @SidharthBansal So I was able to test basic add a step functionality 
test file  
```
const timeout = process.env.SLOWMO ? 30000 : 10000;

beforeAll(async () => {
    await page.goto("http://192.168.43.171:3000/examples/#steps=", {waitUntil: 'domcontentloaded'});
});

describe('title of the page', () => {
    test('Title of the page', async () => {
        const title = await page.title();
        expect(title).toBe('Image Sequencer');
        
    }, timeout);
});
describe('Add step', () => {
    test('length is increased', async () => {
    	await page.waitForSelector('.step');
        const previousLength = await page.evaluate(() => document.querySelectorAll('.step').length);
        await page.click("[data-value='brightness']")
        const previousLength1 = await page.evaluate(() => document.querySelectorAll('.step').length);
        console.log(previousLength)
        console.log(previousLength1)

        expect(previousLength).toBe(1);
        expect(previousLength1).toBe(2);
   }, timeout);
});
```

output file 
```
> image-sequencer@3.5.1 test /home/keshav/image-sequencer
> jest

 PASS  src/test/frontend.test.js
  title of the page
    ✓ Title of the page (10ms)
  Add step
    ✓ length is increased (2269ms)

  console.log src/test/frontend.test.js:20
    1

  console.log src/test/frontend.test.js:21
    2

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        3.804s, estimated 4s
Ran all test suites.

```

I have done it with the jest puppeteer . I found this as the most suitable here. So we won't require ant HTML fixtures as all the things would be done in headless chrome. What are your view's on this

**Comment by jywarren** [comment]
Omg can you share a PR? This would be amazing!!!!!!!!

**Comment by jywarren** [comment]
Solved in #1366 !

**Comment by vaibhavmatta** [comment]
Join me in Learning Competitive Programming Course today

Hello!

I am upskilling myself this summer by doing Competitive Programming Course. As a student of Coding Ninjas, I can invite a few friends to upskill with me this summer.

That said, I and you will get a discount of Rs. 1000 on our courses.

If you are interested, *use this link* ( https://u14128713.ct.sendgrid.net/ls/click?upn=hKenGOKOIZIV7SlorwxvFxg9T9sj5Do2CkUnET1sasw-2Fj07piucWFrdqpoOsvzZU8ID60rSAyZ3b4mSgDVifbx1qsoWKlzrXbfyfusX-2Fw1svpKGtyn3yFTJW7mKYLxqMfwwCOLYe5cuvczrKVnffK9DESCFGgGPfiLPd9HcfJ1bWlpQBb8S6Q7OgcO8rRVO8Tnwqy20-2BUojMzWc6yk-2FskS-2FOYPAM-2B3A4-2B-2B9QQnLXsZ0-3DtfT8_U31jt5HgHBh7k3Grc5s4l2XNXNsg6X5u-2B7fNPeTbqYywXgQbNUrH7o0s00NzOobmaxaqvELQWd8Pacotc7DLz-2BmUpQUOVTWE73NXovvTQp8ABwk1emMI14EIsG-2FqmHlz3AR8IDDHMLMKutlB5vL46bsNLq2BzxxEXJ7zUKTlqZ-2F7UqO26iGvWtl5e8KjuwLz6d1WliUy7WWBZCbzZSwI51mWDAgqHQRZ-2BVK7Oyl4bBBIOHJx-2Fjqxat0Yu03yihZu9i8RlHGr7bIjMMTvw1Nsrqg6D2hWbOC-2BExXtzIXDJW-2BhnRonSvNUTIs0PR2t0YWc-2B7pVXv2-2F7wIxmVScneHeMJ3JqG279d4-2BujoUhRIdkTfZiESHF6H8LaZnW4Cog7W62Mt3MB-2F2vpT-2BkWLJEUfNxxveoGKB0bvkQ-2BN9pB1uNkxpYjGFU1t6ThMat4Mw-2F7XB1LAXzBc3sVp5UfZCA6j-2Biy7ZmixN9Gy7oNBuy5EKsjcw0UYlhD-2FBfWq2Hw6J-2FdHFK-2B0hg-2BrQn7dy3Jq5jEnnX94v0ram6nB0ZOOFi-2BBKLMVR-2FsT-2FY5AWxm3tv9yNe7uPLMl0yFv006f-2FCXtqeM3y8Q-3D-3D ) to start learning on Coding Ninjas.
Unsubscribe From This List https://u14128713.ct.sendgrid.net/asm/unsubscribe/?user_id=14128713&data=TT0P5bbh9Mzz9uBZ1rZ3M0i99Jodfb0nqu0XENx1VgvijxFZ_VRxZGgsvy2fkMuQqVPAz3cVkeeIV2l5lDvrE92ts1jT94fC65yIX_GCJdh_cazWLQ8kyc53XfLNQBZIbndwoLGvBYw_gAD4hXZprwkfzs8IIP-CodVG7nXOWJUwIT6f-b4N0jEFRuidkLozW87T5V97GZ9AdwC8eQmIwZ-iRVNVc1xXY3dnMu1-VmNIS93g5Tz6x_I3K-PAo_am2zqagnu4b0HQTtTAmlq31slIHiJnhBypDXe58eR0KxvPLNU1rtCfpvVpa-sbHn9eoEAGaLmQR7cpk_OPud8ZWBpiX1ZlEIZ_3FUROBwF3LghgyAdaC1qM20-ONs7FRfKzHk_Bn-PNN3mXg0MdC4u09ED1S8FreHRq47z8Gcqy86zWzRbVtbJwOutCYxMmAZ17I4j3iPzgzvwfJ3SZqXj55BXPZenhYcwwNZhhqg_MgFcWOgT2oaRZgE_vXs82kv4AXM_E8G48SeNaXcI4KbP66lMKuXcUJ1EP0BLwe5wHhYL0kQTAK3oP5KxdWxzZE0MrCnrCsshYUgq9hwpl_Iwduy8RJMj1vEbue-Bsl5etuoQr4wk-5upc0tp3Sei7goUhxZLe71VhteDfrXuY5hGW3BArgXgdSoafJABHAJy8MG0x__XWlvgrWMr0C4uWoCa9_N90wkW2l0uqSxYYHWdtg== | Manage Email Preferences https://u14128713.ct.sendgrid.net/asm/?user_id=14128713&data=Bj-nv0TC_Gh0droUFkqbYh0AjMplPbXf6Y3Xk1ViicfIJGjxjuZzdy18LcXZ0XJNa72gmGOuoHAKXVJ4SMEL4pQpS-YD_lR9ACit0Ixt2fKdP0_39aVLOOg_nYwEVyjbQwt3S8quTL2q4GLRRm1UJfnC9mJ_uUsU8SxE8QXBQfSJo2CiGUtR9Vww2pv-Z_aZYn4ARxDV1S9C2LyWxFWdBgsIuUZYLmOvt7TZbKYR8DyAgblrRYjDp3eDi7-FH-GznJj9xDyrJIeNja3bHF_w78NVBC5u3Vep-8mAtzvEJC7MXMMyer8JOdyJtOPOJBEBg09Ne4y7UDl1zJRnSrARbwFNSajmibS-ohMCo5s_cufNbCV1x1nXwUbV01MhRvKKEZYNS1L_buZgCHFq89YkgNvMw5RJHG76i9YQmf6vQ4rXCEznAEzagYTU-5iAO2ccG5DFtt39TVxTvHmX0eBsyTXuOwKYPSPiEVcJy9p-29eJLPd-XwQ9Wk47J5jzuv39ewdOt_NcL30nIwFcjnVjfeUHQ2JG2eBrX8oV3ecvaoqHy8r0UczzjXiGcN-H4LCVv26knX7Fsh-VdVFpWzv8nFJB-TXbnuUdN6AHKX46aNuExtKYExssY20-V9mfTLlRteg8blcg_gqedChhPo_FciqhYQtxbDPcr5JP-pE2-eR4-l9c5ppm1QUmQ9IQyHkpfda_P2usfZD0r5eeJuk6Yg==

**Comment by Divy123** [comment]
Seriously, @vaibhavmatta ?? Is this the right platform for this?

**Pull request closed without merging**